### PR TITLE
Include exception causes

### LINKED ISSUE #7: "Via" information from exception is lost
## Expected behavior

The `:via` information in an exception is made available to Honeybadger.

Honeybadger [supports nested exceptions for Ruby](http://blog.honeybadger.io/honeybadger-now-supports-causes-nested-exceptions/). I'm not sure whether how this should be done via the [`notify` API](http://docs.honeybadger.io/guides/api.html#sample-payload).

## Actual behavior

`:via` information is lost.

## Steps to reproduce

The `:via` information is available after `normalize->event`:

```clojure
(let [e (Exception. "Exception with cause" (Exception. "Cause"))]
  (clojure.pprint/pprint
   (normalize->event "api-key" "env" e {})))

;; => 

{:metadata ...,
 :api-key "api-key",
 :env :env,
 :exception #error {:cause "Cause"
                    :via
                    [{:type java.lang.Exception
                      :message "Exception with cause"
                      :at ...}
                     {:type java.lang.Exception
                      :message "Cause"
                      :at ...}]
                    :trace ...}}
```

but elided by `event->notice`:

```clojure
{:notifier ... ,
 :server ... ,
 :error {:message "Exception with cause",
         :class "java.lang.Exception",
         :backtrace ... ,
         :tags #{}},
 :request ...}
```

### PR DESCRIPTION
**Pull request by xiongtx**
Fix #7.

HB [supports causes](http://blog.honeybadger.io/honeybadger-now-supports-causes-nested-exceptions/). See the [updated documentation](http://docs.honeybadger.io/guides/api.html#sample-payload) on including causes in the payload.

Example:

image

### CODE REVIEW

**Comment by camdez** [comment]
@xiongtx Thank you for your work on this! I actually added support for *a* cause exactly a week ago and wasn't sure how to handle nested causes gracefully. Looks like you have the answer!

Give me just a bit to look at over and I'll get it merged in!

File: src/honeybadger/core.clj [hunk 1]
```diff
@@ -28,6 +28,13 @@
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
+(defn ex-chain
+  "Return all exceptions in the exception chain."
+  [e]
+  (->> e
+       (iterate #(try (.getCause %) (catch NullPointerException _)))
```
Comment (camdez) [inline src/honeybadger/core.clj]: Is there a need for this that can't be obviated by simply checking if the value is `nil`? (I'd rather not cause and catch exceptions in non-exceptional circumstances.)

File: src/honeybadger/core.clj [hunk 2]
```diff
@@ -37,15 +44,23 @@
   (->> st st/parse-trace-elems (map format-stacktrace-elem)))
 
 (defprotocol Notifiable
+  (error-map-no-causes [this])
```
Comment (camdez) [inline src/honeybadger/core.clj]: Do you see a compelling reason to expand the protocol? This seems like just a concern of `Throwable`'s implementation of the protocol, not the protocol itself.

File: src/honeybadger/core.clj [hunk 3]
```diff
@@ -37,15 +44,23 @@
...
-    (error-map [this] {:message (str this)
-                       :class (.getName (.getClass this))
-                       :backtrace (format-stacktrace (.getStackTrace this))}))
+  (error-map-no-causes [this]
+    {:message (str this)
+     :class (.getName (.getClass this))
+     :backtrace (format-stacktrace (.getStackTrace this))})
+  (error-map [this]
+    (let [exs (rest (ex-chain this))]
+      (merge (error-map-no-causes this)
```
Comment (camdez) [inline src/honeybadger/core.clj]: This could be simplified to an `assoc`.

**Review by camdez — COMMENTED** [review]
LMK your thoughts on these comments and we'll work towards getting this merged.

File: src/honeybadger/core.clj (hunk 1 above)
Comment (xiongtx) [reply src/honeybadger/core.clj]: :white_check_mark:

File: src/honeybadger/core.clj (hunk 2 above)
Comment (xiongtx) [reply src/honeybadger/core.clj]: :white_check_mark:

File: src/honeybadger/core.clj (hunk 3 above)
Comment (xiongtx) [reply src/honeybadger/core.clj]: :white_check_mark:

**Comment by xiongtx** [comment]
Addressed comments :ok_hand:.

> Looks like you have the answer!

The folks at @honeybadger-io kindly updated the documentation so I could figure out what's going on :clap:.

**Comment by camdez** [comment]
@xiongtx Looking good! Let me do a little testing and then I'll get it merged / cut a new release. (Also gonna rebase to `dev`. I've only been merging to `master` for release.)

> The folks at @honeybadger-io kindly updated the documentation so I could figure out what's going on 👏.

Thanks, kind @honeybadger-io folks! 🎉

**Comment by camdez** [comment]
Merged! (Though apparently not in a way that made GitHub properly realize it was merged. :/ )

I've pushed a snapshot (pre-release) version to Clojars: https://clojars.org/camdez/honeybadger/versions/0.3.1-SNAPSHOT

I'm going to run that in one of my own production environments for a day or two just to make sure we're solid, and then I'll ship a proper 0.4.0 release.

Thanks again for your hard work! 🥂

**Comment by xiongtx** [comment]
> I'm going to run that in one of my own production environments for a day or two 

Sounds good, but it'd be nice if there were better tests. We can then have greater confidence that changes have't introduced regressions.

One thing I'd like to see is replacing `schema` w/ `spec` and specing out a [notice](https://github.com/honeybadger-io/honeybadger-ruby/blob/master/lib/honeybadger/notice.rb).

**Comment by camdez** [comment]
@xiongtx completely agree on both. PRs welcome. 😜 Or I’ll eventually find the time. 

FWIW I did add a quick test for these changes, but the project obviously needs more.

**Pull request closed without merging**